### PR TITLE
Draft proposal for hub permissioning system

### PR DESCRIPTION
**Pull request by thedoctor**
This PR includes a plain-english proposal for an access control system for hubs.

Major points:
- recommends CRUDX as the permission level set
- ~removes factors (which AFAICT can be encapsulated with DIDs)~
- considers all paths equivalent for the purpose of access control (i.e. no special treatment for built-in routes) with the exception of /stores, which will use the same verification system, but whose permissions will be hard-coded.
- suggests a inheritance model where a permission specification can be applied to fields inside data objects EDIT: marked for future inclusion
- ~deliberately does not address a callback system because I think that should probably be handled elsewhere~
- ~fails to address querying/tagging/content-types because that raises broader questions that have implications for the entire system which need to be discussed~
  Expands the previous callback system to support URIs, headers and other fields
- introduces an `ext` field for route-specific permissions

Minor points:
- the sub-header markdown links probably don't work

### CODE REVIEW

**Comment by csuwildcat** [comment]
Feedback and questions:

- I like the fact that each permission object is itself a schema'd JSON LD obj
- `/stores` is a route that is only permissioned in one way: full control to the identity owner and external identity a store is granted to, with no variability - think of this as the private `window.localStorage` of Hubs
- Maybe I missed it, but is there included in this the concept of permissioning for just the things the external identity creates vs those created by all other identities, including the owner?
- Can you elaborate a bit more on the inheritance side of things? Is this related to nested schema objects? We want to make sure that collections for a schema are always flattened into a single depth surface, so levels of inheritance shouldn't apply there - was this what you meant?

**Comment by thedoctor** [comment]
@csuwildcat 
> /stores is a route that is only permissioned in one way: full control to the identity owner and external identity a store is granted to, with no variability - think of this as the private window.localStorage of Hubs
I forgot about this, but this raises two questions for me:
- Will there ever be a situation when the owner wants to allow multiple identities to share a resource posted to its store? Is the way to address this copying it to a collection? Should, then, there be a proxy collection type that simply routes to a store?
- Either way, we can still represent the desired behavior using the proposed system. I would argue that it's better to use a consistent mechanism rather than to build in special behavior for one route (especially if '/stores' is the only route that wouldn't be subject to the normal permission system)

> ... is there included ... the concept of permissioning for just the things the external identity creates vs those created by all other identities ...
Not explicitly, but that can be represented here and the guts of constructing those permissionspecs can be hidden. I wasn't thinking of this case though, so maybe there's a simpler way of supporting this case.

> Can you elaborate a bit more on the inheritance side of things?
So I probably missed this somewhere in the explainer, but I wasn't under the impression that a collections are 1:1 with schemas, I was thinking of them as arbitrarily-namespaced directories. If we're not worried about supporting hierarchies, then most of the inheritance stuff and some of the tag/query stuff doesn't really apply –– UNLESS we want to be able to independently permission _fields_ within a document (which we do at Gem; a simple use case being a medical record where PII is visible only to the patient and their primary physician, but not to analysts doing e.g. ML on lots of medical records)

**Comment by TelegramSam** [comment]
Your proposal is a great beginning.

Is there a sane way to designate sub-path permissions? For example, can we allow both a full and a subset of the /profile data to be returned?

Or, on a collection, can I show you my playlist, but not the play counter for each song, present in the same collection?

**Comment by thedoctor** [comment]
@TelegramSam So yeah, this goes back to the independent permissioning of fields within a document, inheritance, and pattern matching choices. Example:

SimplePerson schema:
```json
{
  "@id": "myhub.io/collections/SimplePerson",
  "first_name": {"type": "string"},
  "last_name": {"type": "string"}
}
```

To grant Sam permission to only `first_name`s, you could add this block to your permission set:
```json
{
  "identifiers": [ "did:sov:sam.id" ],
  "path": "myhub.io/collections/SimplePerson/*/#first_name",
  "allow": "-R---"
}
```

OR you could add both of these:
```json
[{
  "identifiers": [ "did:sov:sam.id" ],
  "path": "myhub.io/collections/SimplePerson/*",
  "allow": "-R---"
}, {
  "identifiers": [ "did:sov:sam.id" ],
  "path": "myhub.io/collections/SimplePerson/*/#last_name",
  "deny": "-R---"
}]
```
In this example schema, those would be equivalent. When `did:sov:sam.id` fetches `myhub.io/collections/SimplePerson/id-of-mattsmith` he would only see:
```json
{
  "@id": "myhub.io/collections/SimplePerson/id-of-mattsmith",
  "first_name": "matt"
}
```

**Comment by csuwildcat** [comment]
In terms of property-level permissioning within objects, I am weary to open that Pandora's Box for many reasons. Objects may be encrypted for user-only and whitelisted access, and the host of the Hub is not privy to the data. This means You would need to devise a scheme of complex property management and encryption that may even require duplicating property-level data for different groups.

This is something I'd really like to avoid for v1, if possible, because it makes the permissioning and implementation much more difficult, imo.

**Comment by csuwildcat** [comment]
As for `/stores`, the reason I would hard code the security model and not have it be added as some default permission are two fold:

1. It is specifically designed to be a 1:1 relationship with the owner and the external party. Imagine if in a browser's code base someone told you "window.localStorage is only for the user and the domain to access, and that will literally never change, ever, forevah evah, but we made it a configuration so that we could technically expose it."
2. On the security front. the best way to close an attack surface is to never create one. WRT permission systems, wherever you can remove the _ability_ to permission altogether, you should do so, because it is just another open door for compromise and possible implementer mistakes.

**Comment by csuwildcat** [comment]
WRT the `factors` section of the permissioning scheme:

This was intended to be devices and machines that you own that are control devices for your identity, not different identities. The implication here is that they are wholly different, as they are not registered as DIDs (because they're not identities). Essentially, these are connected devices/hardware factors that _are_ your identity access points. The big issue here is that you'd have to have really complex management of an enormous whitelist that always was a 100% super set of all data paths the Hub holds. This could be extremely difficult, error prone code, imo.

**Comment by csuwildcat** [comment]
On callbacks:

This is a fundamental need in the system, which many app/service cases will rely on, so I would like to see us incorporate it. This strikes me as one of the least difficult areas of the permission system.

**Comment by csuwildcat** [comment]
For querying and tagging:

We could elect to wrap every object added to the system in a control object. This would allow us to keep all system properties and operational data out of the pure data objects. Some of these properties could include things like tags, metadata, and other things that we want to allow querying of. The necessity for this is that the Hub host may not be trusted with a source object's raw data, so tags/metadata would need to exist outside the encrypted object.

**Comment by csuwildcat** [comment]
On the device/factor front: we could seek to add those as permissioned entities in the regular flow if we want varied levels of permissions. I'd think we need to add a couple flags that indicate the device is a sync'd factor, to facilitate change feed access and other things required for these special devices/factors.

**Comment by TelegramSam** [comment]
re: field-level permissions. I think it may be worth shelving that one for now, but keeping it in mind. Even in /profile, having a basic profile and an extended profile would be very useful. We could get around that with sub-routes that are permissioned. As far as encrypted data: that should be a black box. Field level filters cannot take effect inside an encrypted block. This doesn't prevent the filters from working on non-encrypted data.

**Comment by csuwildcat** [comment]
@TelegramSam I agree - per-field gets a hard really quick; and we could likely add it in the future, if there was a strong need expressed by the community.

That said, maybe we special-case `/profile` to allow for a different object composition based on the inbound ID that is requested. Here's what I mean:

Imagine I have a DID that maps to 2 names,`dan.id` and `anonguy.id`, and I wanted each to display a different profile to the world. The inbound route will contain the ID the requesting party is asking for, thus the Hub that knows of the two pseudonyms could output a different object based on which is requested - make sense?

To do this, we could create a permission section that allowed identity owners to specify IDs/names mapped to profile object UIDs or paths.

Thoughts?

**Comment by thedoctor** [comment]
@TelegramSam @csuwildcat 

RE field-level permissions: Perfectly reasonable to punt this for later, as long as we know our permissioning system will be capable of supporting it in a straightforward way (which it is as of the current state of this PR). IMO, encrypted files won't really affect this; the hub simply can't enforce access controls against something to which it doesn't have access (`GET /hub.io/collections/C/encrypted_document_id/present_but_encrypted_field` would just be a 404), so I don't see a problem there. 

RE `/profile`: I think it definitely makes sense to support per-DID profiles, and I agree we should be able to permission those independently, but I wouldn't actually call that a special case. I suggest that a DID's profile lives at `HUB/profile/dan.id`/`HUB/profile/anonguy.id`/etc and those routes are permissioned the same way as everything else. Finally there does need to be a configuration mechanism for routing from `/profile` to a default DID's profile, and that could possibly be a function of the identity making the incoming request, but I think that is outside the scope of the permissions system. Hopefully this isn't controversial.
What may be a little controversial is that I don't think we should construct those profile objects dynamically, they should be concrete documents and we should simply map DIDs to complete profile documents.

**Comment by csuwildcat** [comment]
I agree with what you wrote @thedoctor, with one note: the route structure would be `/.identity/dan.id/profile` and `/.identity/anonguy.id/profile`. (I just wasn't sure if that was a transposition or intended)

**Comment by thedoctor** [comment]
@csuwildcat RE factors:
> The big issue here is that you'd have to have really complex management of an enormous whitelist that always was a 100% super set of all data paths the Hub holds. This could be extremely difficult, error prone code, imo.

I don't understand your concern. My understanding is that the DID spec gives us all the flexibility we need to say `matt's DID is controlled by such-and-such a quorum of these (other DIDs/public keys)`. Why duplicate that internally in hubs? I would argue that anything incapable of holding a key and making a signature isn't an acceptable identity factor, and if it can do those things, why not just represent it as a DID since it would be useful to other systems that need to authenticate an identity as well? 
In fact, it seems to me that representing factors as DIDs would simplify a given hub's permission document, rather than complect it.

I think this goes back to the philosophical argument of whether a cell phone is an identity or not. Historically, I've been of the opinion that `things` which act on-behalf or at-behest of `individuals-or-organizations` aren't actually identities, but I think I'm in the minority there. But even if everyone agreed that a phone isn't an identity, I _still_ don't see a reason for hub to implement factors since we could still just make `matt.id`'s DID control group contain the phone's key in an m-of-n configuration.

TL;DR seems like reinventing the wheel any way I think about it.

**Comment by thedoctor** [comment]
@csuwildcat RE `/stores`:
Okay, I see your point, I'm fine saying that store permissions are fixed, but to achieve that we can just hardcode the permissions for a store to be `OR(store's external identity, identity owner)`, full stop. The permission system can still be responsible for parsing that and validating incoming requests against it, don't you think? I would argue that re-implementing a subset of that logic specifically for the stores endpoint expands the attack surface for no benefit apart from security theatre. Maybe we're in agreement there?

Other than that, I have no objections except that it means I have to duplicate anything that gets put into `/stores` that I want to re-use or expose to anyone else. I don't have any use cases that require that, so I can't defend it as an essential feature.

**Comment by thedoctor** [comment]
@csuwildcat RE callbacks:
> This is a fundamental need in the system
100% agreed

> This strikes me as one of the least difficult areas of the permission system
So what I was saying was that I don't see how this relates to the permission system beyond that an external identity should only be able to be notified about things they have permission to access in the first place. We definitely need callbacks, but I don't think they are part of the access control subsystem.
Maybe you could explain your vision of the callback interface a little more. Were you imagining a separate "callback permission" or something?
I'm imagining a mechanism where external actors provide
1. some sort of a query (depending on what kind of query system we support) and 
2. a callback URI (pointing to either a hub or a normal webserver)
and will receive an HTTP POST request every time a datum is added or modified such that it matches the query _and is permitted to be READ by the external identity_.

**Comment by thedoctor** [comment]
And finally RE querying:

> We could elect to wrap every object added to the system in a control object. 

+1 works for me, but 2 things:
1. I think it would be reasonable to have a control object that signals to the hub to allow querying against the actual data object if 
    - the hub _does_ has access to it at runtime 
    - its content is structured in a supported way (json/yaml/xml)
    - the querier is authorized
2. We still need to figure out the actual query DSL

**Comment by csuwildcat** [comment]
One major issue with forcing every factor/device the user owns to be a DID (separate identity) or an update to the DDO values, is that it will (in many cases) require an update transaction on the chain of the ID root. This means that adding a device could take hours, or longer, for some chains.

By moving this to the Hubs only, and leveraging the chain for anchored trust of the signature on the entry, we can add devices, propagate the change to all Hubs, and enable use in milliseconds, not hours.

**Comment by csuwildcat** [comment]
WRT `/stores`, I definitely would reuse the identity validation and processing mechanisms for this API area, but would caution it as a mutable or pre-populated ACL entry that just stays static - we hope. It's better for sec that this is never possible to change from a config entry, because it eliminates another possible vector for breaking what amounts to a cross-origin Web boundary (to borrow a similar concept from browser land).

**Comment by thedoctor** [comment]
> would caution it as a mutable or pre-populated ACL entry that just stays static - we hope
Totally, I'm on board – it shouldn't be an entry, the store's permission should get calculated every time based solely on its path (e.g. `hub/stores/dan.id`) and shouldn't be mutable or even persisted.

**Comment by csuwildcat** [comment]
When you list the paths like this: `hub/stores/dan.id`, do you intend that sequence, or do you mean `hub/dan.id/stores/spotify.id`?

**Comment by thedoctor** [comment]
In the previous example I was accidentally transposing. In this case I meant:
`MATT'S-HUB-ROOT-PATH/stores/dan.id` which would be dan's store in matt's hub. Ambiguity acknowledged. From now on I'll stick to using `HUB/identity-owner.id/api-path` for clarity

**Comment by csuwildcat** [comment]
> Maybe you could explain your vision of the callback interface a little more. Were you imagining a separate "callback permission" or something? I'm imagining a mechanism where external actors provide some sort of a query (depending on what kind of query system we support) and a callback URI (pointing to either a hub or a normal webserver) and will receive an HTTP POST request every time a datum is added or modified such that it matches the query and is permitted to be READ by the external identity.

Here's what it would look like to give Dr John permissions to all my HL7 FHIR medical data, and some callbacks associated with that grant - in a format that's a bit of a combination of your proposal and other feedback:

```json
{
"did": "doctor_john.id",
  "stores": {
    "max_size": "100"
  },
  "collections": [
    {
      "identifiers": ["dan.id", "secret_patient.id"],
      "path": "hl7.org:fhir/*",
      "allow": "-R---",
      "callbacks": [ "created", "modified" ]
    }
  ]
}
```
Here's what's happening above:

- The DID being granted permissions is `doctor_john.id`, as indicated by the `did` field on the base of his specific permission object
- The presence of a `stores` field indicates Dr. John is being given a JSON storage space in the Hub, with the max_size set to 100, which is always in megabytes and can be a float
- the collections permissions are in the format you describe in your doc, but the identifiers are about _your_ DID/name pseudonyms, not Dr. Johns - not sure if this was clear from the explainer.
- The `path` is assumed to begin at collections
- The `allow` follows your CRUDX proposal
- For `callbacks`, the entry causes the Hub provider to notify the Hub of `doctor_john.id` when something happens matching a granted callback type. In this case, a schema.org [`Message`](http://schema.org/Message) object with  `potentialAction` object of of [`UpdateAction`](http://schema.org/UpdateAction) is sent to the Dr. John's '/messages' Hub endpoint whenever a FHIR object is created or modified in the user's Hub.

Thoughts?

**Comment by csuwildcat** [comment]
I think the top-level 'native' routes are different enough that they warrant their own subsections in permission objects that contain the aggregate permissions of the permitted outside identity, which is show in the merged format example above.

**Comment by TelegramSam** [comment]
On different profile responses per caller: I expect this to be common practice, though I suspect we are thinking about it in different ways.

I envision a hub that has multiple disparate (and un-correlatable) identifiers. Each DID points to its own unique hub identifier. The hub considers the inbound hub identifier during the response and returns the proper information based on that context.

I know your hub vision revolves around publicly known hub identifiers. We need to ferret this out so that it works for both cases.

An open question in my mind is what falls inside the hub spec, and what is left to the discretion of the implementer. Determining the proper profile response may fall into that category.

**Comment by TelegramSam** [comment]
I agree on @thedoctor's last statement on /stores permissions. This is simple and doesn't need to be part of the rest of the system.

**Comment by csuwildcat** [comment]
> I know your hub vision revolves around publicly known hub identifiers. We need to ferret this out so that it works for both cases.

> The hub considers the inbound hub identifier during the response and returns the proper information based on that context.

How are our proposals different? I actually thought we were aligned on this point and the proposed implementation. No?

**Comment by csuwildcat** [comment]
@TelegramSam can you provide some further detail on this:

> I agree on @thedoctor's last statement on /stores permissions. This is simple and doesn't need to be part of the rest of the system.

I am not sure what you meant.

**Comment by TelegramSam** [comment]
@csuwildcat on my stores comment. I was agreeing with @thedoctor here: https://github.com/decentralized-identity/hubs/pull/18#issuecomment-314238127

**Comment by csuwildcat** [comment]
@telegramsam makes sense - we just handle the access aspect of stores in the module itself (reusing any identity validation check code).

**Comment by TelegramSam** [comment]
We might not view this as differently as I initially though. Let me try and unpack this:
> Imagine I have a DID that maps to 2 names,dan.id and anonguy.id, and I wanted each to display a different profile to the world. 

I would read this as 2 names (dan.id and anonguy.id) that map to the same DID. The DDO for that DID contains a service pointer to a hub.

>The inbound route will contain the ID the requesting party is asking for, thus the Hub that knows of the two pseudonyms could output a different object based on which is requested - make sense?

The inbound route would be identical in this case (the one specified in the DDO). The hub can still return differing information based on the identity of the requesting party, but those responses are now correlatable because they came from the same DID/DDO/Hub.

A solution which avoids the correlation problem looks like this:
- dan.id and anonguy.id each resolve to a different DID. Each DID has it's own DDO, and each has a unique hub service pointer.
- Both unique hub URIs resolve to the same logical hub by the hub host. This allows one logical hub to serve requests for both (virtual?) hubs.
- The hub host provides the virtual route to the hub during the request processing. The hub then can differ the response by which URI was used.
- As long as the response contains no correlatable information between the responses for dan.id and anonguy.id, no correlation can be made.

In the Sovrin pairwise identifier world, each hub will serve hundreds, if not thousands, of unique HUB URIs that resolve to the same logical hub. I know we are unique in this: most other solutions rely more heavily on well-known identifiers.

**Comment by TelegramSam** [comment]
On 3rd party use of the /permissions endpoint:

A 3rd party requesting the /permissions endpoint would only be returned the permissions they've been granted, right? The owner would get them all of course.

Does a 3rd party request permissions by posting a new object to this endpoint? How does the workflow of this work?

**Comment by csuwildcat** [comment]
@TelegramSam we are very close - all we need to do to add pairwise, parallel, (internally) Hub associated TLNs/DIDs for the same core identity is to provide a call/route under `/permissions` that allows an identity owner to verify ownership of additional TLNs/DIDs. Hubs can then manage them as one entity with many personas, depending on the querying party and other contextual info. The identifier filtering permissioning strategy I proposed would not have to change, we just need to add the functionality I mentioned.

**Comment by csuwildcat** [comment]
Question: did we ever make a call on device/factor associations? If we say that factors/devices must be associated via DDOs, it would mean you need to do a chain transaction to updated the DDO. If you do it via a direct write of a Hub data/permission entry of some kind, you can instantly add/remove new devices.

I'd love to hear your collective thoughts on this, as it will be immediately required for most use cases.

**Comment by thedoctor** [comment]
Ah yeah, I'm not sure if we landed on a decision – it may have slipped through the cracks, but I agree that it's unacceptable to have to e.g. wait for a block to revoke access to a lost device.

IIRC factors are replicated between hubs, i.e. they're local to an identity owner, not local to a hub? Or ought they support both modes?

**Comment by csuwildcat** [comment]
> IIRC factors are replicated between hubs, i.e. they're local to an identity owner, not local to a hub? Or ought they support both modes?

I didn't understand this, can you explain a bit more about what you mean by local/non-local?

**Comment by thedoctor** [comment]
Like, if I add a factor to a hub, will that factor be valid for connecting to my other hubs?

**Comment by csuwildcat** [comment]
@thedoctor yes, it would be a replicated entry just like other permissions, data objects, and messages. The entry would be signed by a higher-order key or form of DID-linked proof that basically informs the Hubs: "Trust data coming in signed with this public attestation key as me."

**Comment by thedoctor** [comment]
So questions:
- Factors create an off-chain association between a public key and the identity owner (not a specific DID)? 
- All factors must minimally be capable of signing an HTTP request?

**Comment by csuwildcat** [comment]
> Factors create an off-chain association between a public key and the identity owner (not a specific DID)?

Yes, a device/factor is a signed attestation from the identity owner that payloads coming through from the device/factor are to be regarded as from the identity owner.

> All factors must minimally be capable of signing an HTTP request?

Not sure - I suppose you could have a factor that holds a key and is used by some other random device that performs the actual request?

**Comment by thedoctor** [comment]
Whoops closed accidentally.

> I suppose you could have a factor that holds a key and is used by some other random device that performs the actual request?
Yeah I was thinking like a trezor wallet — but I didn't think that was contentious, I was clarifying that we don't need to support factors that might not have crypto operation support.
I think we're in agreement, I'm fine with including factors as specified in the original explainer. I'll add them back in this pr when I get to a computer. I'll make a separate authentication.md.
